src/sql/workbench/api/node/extHostNotebookDocumentsAndEditors.ts

```diff
@@ -152,6 +162,33 @@ export class ExtHostNotebookDocumentsAndEditors implements ExtHostNotebookDocume
 		}
 	}
 
+	private _nextHandle(): number {
+		return ExtHostNotebookDocumentsAndEditors._handlePool++;
+	}
+
+	private _addNewAdapter(adapter: Adapter): number {
+		const handle = this._nextHandle();
+		this._adapters.set(handle, adapter);
+		return handle;
+	}
+
+	private _getAdapter<T>(id: number): T {
+		let adapter = <T><any>this._adapters.get(id);
+		if (adapter === undefined) {
+			throw new Error('No adapter found');
+		}
+		return adapter;
+	}
+
+	$getNavigation(handle: number, notebookUri: UriComponents): Thenable<azdata.nb.NavigationResult> {
+		let navProvider = this._getAdapter<azdata.nb.NavigationProvider>(handle);
+		if (navProvider) {
+			let uri = URI.revive(notebookUri);
+			return navProvider.getNavigation(uri);
+		}
+		throw new Error('No navigation provider found for handle ' + handle);
```

Comment:
>'No navigation provider found for handle ' + handle [](start = 18, length = 51)

nit: can just use externalized string here - makes it slightly less verbose and easier to read

`No navigvation provider found for handle ${handle}`